src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -4414,6 +4414,216 @@ class StubGenerator: public StubCodeGenerator {
+    __ srli(tmp1, R_0, 2);
+    __ slli(tmp2, tmp1, 2);
+    __ add(RR_0, tmp1, tmp2);
+    __ srli(tmp1, R_1, 2);
+    __ slli(tmp2, tmp1, 2);
+    __ add(RR_1, tmp1, tmp2);
+
+    // U_n is the current checksum
+    const Register U_0 = *++regs, U_1 = *++regs, U_2 = *++regs;
+    pack_26(U_0, U_1, U_2, acc_start);
+
+    static constexpr int BLOCK_LENGTH = 16;
+    Label DONE, LOOP;
+
+    __ addi(tmp1, zr, checked_cast<u1>(BLOCK_LENGTH));
+    __ blt(length, tmp1, DONE); {
+      __ bind(LOOP);
+
+      // S_n is to be the sum of U_n and the next block of data
+      const Register S_0 = *++regs, S_1 = *++regs, S_2 = *++regs;
+      __ ld(S_0, Address(input_start, 0));
+      __ ld(S_1, Address(input_start, wordSize));
+
+      __ cad(S_0, S_0, U_0, tmp1); // Add U_0 to S_0 with carry output to tmp1
+      __ cadc(S_1, S_1, U_1, tmp1); // Add U_1 with carry to S_1 with carry output to tmp1
+      __ adc(S_2, zr, U_2, tmp1);
+
+      __ addi(S_2, S_2, 1);
+
+      const Register U_0HI = *++regs, U_1HI = *++regs;
+
+      // NB: this logic depends on some of the special properties of
+      // Poly1305 keys. In particular, because we know that the top
+      // four bits of R_0 and R_1 are zero, we can add together
+      // partial products without any risk of needing to propagate a
+      // carry out.
+      wide_mul(U_0, U_0HI, S_0, R_0);  wide_madd(U_0, U_0HI, S_1, RR_1, tmp1, tmp2); wide_madd(U_0, U_0HI, S_2, RR_0, tmp1, tmp2);
+      wide_mul(U_1, U_1HI, S_0, R_1);  wide_madd(U_1, U_1HI, S_1, R_0, tmp1, tmp2);  wide_madd(U_1, U_1HI, S_2, RR_1, tmp1, tmp2);
+      __ andi(U_2, R_0, 3);
+      __ mul(U_2, S_2, U_2);
+
+      // Recycle registers S_0, S_1, S_2
+      regs = (regs.remaining() + S_0 + S_1 + S_2).begin();
+
+      // Partial reduction mod 2**130 - 5
+      __ cad(U_1, U_1, U_0HI, tmp1); // Add U_0HI to U_1 with carry output to tmp1
+      __ adc(U_2, U_2, U_1HI, tmp1);
+      // Sum now in U_2:U_1:U_0.
+      // Dead: U_0HI, U_1HI.
+      regs = (regs.remaining() + U_0HI + U_1HI).begin();
+
+      // U_2:U_1:U_0: += (U_2 >> 2) * 5
+      __ srli(tmp1, U_2, 2);
+      const int64_t bits2 = right_n_bits(2);
+      __ andi(U_2, U_2, bits2); // Clear U_2 except for the first two bits
+      __ slli(tmp2, tmp1, 2);
+      __ add(tmp1, tmp1, tmp2); // Impossible to overflow since two leftmost bits are zero'ed in 'srli(tmp1, U_2, 2)'
+      __ cad(U_0, U_0, tmp1, tmp3); // Add tmp1 (= (U_2 >> 2) * 5) to U_0 with carry output to tmp3
+      __ cadc(U_1, U_1, zr, tmp3); // Add carry to U_1 with carry output to tmp3
+      __ add(U_2, U_2, tmp3);
+
+      __ sub(length, length, checked_cast<u1>(BLOCK_LENGTH));
+      __ addi(input_start, input_start, 2 * wordSize);
+      __ addi(tmp1, zr, checked_cast<u1>(BLOCK_LENGTH));
+      __ bge(length, tmp1, LOOP);
+    }
+
+    // Further reduce modulo 2^130 - 5
+    __ srli(tmp1, U_2, 2);
+    __ slli(tmp2, tmp1, 2);
+    __ add(tmp1, tmp1, tmp2); // tmp1 = U_2 * 5
```

Comment:
can we replace lines 4581 and 4582 with shadd(tmp1, tmp1, tmp1, tmp2, 2); ?

Reply:
Replaced